crates/store/re_log_encoding/src/rrd/headers.rs

```diff
@@ -0,0 +1,263 @@
+            &[compression, serializer, 0, 0] => {
+                let compression = match compression {
+                    0 => Compression::Off,
+                    1 => Compression::LZ4,
+                    _ => return Err(OptionsError::UnknownCompression(compression).into()),
+                };
+                let serializer = match serializer {
+                    1 => return Err(OptionsError::RemovedMsgPackSerializer.into()),
+                    2 => Serializer::Protobuf,
+                    _ => return Err(OptionsError::UnknownSerializer(serializer).into()),
+                };
+                Ok(Self {
+                    compression,
+                    serializer,
+                })
+            }
+
+            _ => Err(OptionsError::UnknownReservedBytes.into()),
+        }
+    }
+}
+
+// ---
+
+#[derive(Debug, Clone, Copy)]
+pub struct StreamHeader {
+    pub fourcc: [u8; 4],
+    pub version: [u8; 4],
+    pub options: EncodingOptions,
+}
+
+impl StreamHeader {
+    pub const ENCODED_SIZE_BYTES: usize = 12;
+}
+
+impl StreamHeader {
+    pub fn to_version_and_options(
+        self,
+    ) -> Result<(CrateVersion, EncodingOptions), crate::rrd::CodecError> {
+        {
+            // We used 0000 for all .rrd files up until 2023-02-27, post 0.2.0 release:
+            let encoded_version = if self.version == [0, 0, 0, 0] {
+                CrateVersion::new(0, 2, 0)
+            } else {
+                CrateVersion::from_bytes(self.version)
+            };
+
+            if encoded_version.major == 0 && encoded_version.minor < 23 {
+                // We broke compatibility for 0.23 for (hopefully) the last time.
+                return Err(crate::rrd::CodecError::IncompatibleRerunVersion {
+                    file: Box::new(encoded_version),
+                    local: Box::new(CrateVersion::LOCAL),
+                });
+            } else if encoded_version <= CrateVersion::LOCAL {
+                // Loading old files should be fine, and if it is not, the chunk migration in re_sorbet should already log a warning.
+            } else {
+                re_log::warn_once!(
+                    "Found data stream with Rerun version {encoded_version} which is newer than the local Rerun version ({}). This file may contain data that is not compatible with this version of Rerun. Consider updating Rerun.",
+                    CrateVersion::LOCAL
+                );
+            }
+        }
+
+        Ok((CrateVersion::from_bytes(self.version), self.options))
+    }
+}
+
+impl Encodable for StreamHeader {
+    fn to_rrd_bytes(&self, out: &mut Vec<u8>) -> Result<u64, crate::rrd::CodecError> {
+        let Self {
+            fourcc,
+            version,
+            options,
+        } = self;
+
+        let before = out.len() as u64;
+
+        out.extend_from_slice(fourcc);
+        out.extend_from_slice(version);
+        {
+            // TODO(cmc): the extra heap-alloc and copy could be easily avoided.
```

Comment:
why don't you then? :troll:

More seriously, I don't mind these kinds of TODO (culprit of the same), but I do try to leave a breadcrumb as to why I didn't (or open an issue, etc.)